### PR TITLE
Ensure AsyncQueue events have sufficient exception handling

### PR DESCRIPTION
**Pull request by MithrilMan**
refers to #2160 

added OnEnqueueAsyncFails to handle uncatched exception on AsyncQueue callbacks

Not mergeable right now, just pushed to discuss further implementation

With current code now is possible to handle currently unhandled exception (to do)

Do you like this approach or would you like to just check each of the AsyncQueue usage and ensure we produce just a console output/log?

### CODE REVIEW

**Comment by MithrilMan** [comment]
usage example
currently we have this
https://github.com/stratisproject/StratisBitcoinFullNode/blob/664e209a388b8877dc5dc7b881e7cf83caaffe90/src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs#L66

https://github.com/stratisproject/StratisBitcoinFullNode/blob/664e209a388b8877dc5dc7b881e7cf83caaffe90/src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs#L74-L83

this can be replaced by

`this.peersToDispose = new AsyncQueue<INetworkPeer>(this.OnEnqueueAsync, this.OnEnqueueAsyncFails);`

```
private Task<bool> OnEnqueueAsyncFails(INetworkPeer item, CancellationToken cancellationToken, Exception ex)
{
    this.logger.LogError(ex, "OnEnqueueAsync failed");

    // false will cause the AsyncLoop to throw.
    // return true if this error is handled properly and execution can continue.
    return Task.FromResult(false);
}
```

File: src/Stratis.Bitcoin/Utilities/AsyncQueue.cs [hunk 1]
```diff
@@ -147,7 +161,23 @@ private async Task ConsumerAsync()
...
+                        try
+                        {
+                            await this.onEnqueueAsync(item, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (Exception ex)
+                        {
+                            bool exceptionHandled = false;
+                            if (this.onEnqueueAsyncFails != null)
+                            {
+                                exceptionHandled = await this.onEnqueueAsyncFails(item, cancellationToken, ex);
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Utilities/AsyncQueue.cs]: Ah if the exception is handled then the loop will continue? yeah I agree.

**Comment by noescape00** [comment]
I think its a bad idea. 
How should the components use this callback?
Also it wont be possible to dispose the AQ from that callback. 

Also all components that use AQ should never throw in the callback. If they can throw it should be handled on component's level and not thrown into AQ.

**Comment by MithrilMan** [comment]
I forget to push an example on how to use it, but it's pretty straightforward
In this example I do nothing but just log that something happened and return false (so nobody still handle the exception and AQ will throw as before

If you want to handle the exception, you can either do that in the OnEnqueueAsync itself (preferred choice) or catch it in "OnEnqueueAsyncFails" and decide if handle it and return true, or do nothing and return false

The goal was to have evidence on console/log about what was happening, if we had a logger in AQ I could have logged the error before throwing, what do you think about adding logger to AQ constructor?

**Comment by MithrilMan** [comment]
with latest commit I've passed a logger instance along the chain, so AsyncQueue will report a LogError in case an exception happens and there was no onEnqueueAsyncFails handler (so basically it doesn't change current behavior, just add an explicit LogError to evidence there was an exception that stopped the AsyncQueue

**Comment by dangershony** [comment]
>Also all components that use AQ should never throw in the callback. If they can throw it should be handled on component's level and not thrown into AQ.

If they throw in the callback its the same behaviour as we have now.

>I think its a bad idea.

How else will you suggest to handle errors?

**Comment by MithrilMan** [comment]
Now that I've added a logger to AQ, we have a way to spot unhandled AQ exceptions.
To simplify I'll remove the OnEnqueueAsyncFails and will start to implement proper exception handling in the action AQ will run.

Before that I'll list all the usage we currently have of AQ to decide what's the proper way to handle exception for each use case
In some scenario we should probably trigger a NodeLifeTime.StopApplication.

**Comment by MithrilMan** [comment]
Here a list of components that doesn't handle exceptions:

**NetworkPeerDispose:** 
It doesn't have any execption handler

https://github.com/stratisproject/StratisBitcoinFullNode/blob/88363533ce8ce92dc4fcd21a8fef1fcc1e37d91d/src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs#L74-L83
--

**PartialValidator**
Would be better to move first line in try/catch block

https://github.com/stratisproject/StratisBitcoinFullNode/blob/88363533ce8ce92dc4fcd21a8fef1fcc1e37d91d/src/Stratis.Bitcoin/Consensus/Validators/BlockValidator.cs#L144-L157
--

**CallbackMessageListener**  (NetworkPeerConnection)
Probably need some refactoring (could even inherit from AQ).
Actually it's a wrapper around AsyncQueue and it's used by NetworkPeerConnection.
The "problem" is that the action to perform is bassed in its constructor, so it's not easy to know what to do with exceptions

**Comment by noescape00** [comment]
>How else will you suggest to handle errors?

Suppress all exceptions (wrap callbacks in try catch), when exception occurs log it and then carry on or dispose the node based on component's importance.

**Comment by noescape00** [comment]
>NetworkPeerDispose:
>It doesn't have any execption handler

>PartialValidator
>Would be better to move first line in try/catch block

Those can't throw. 
But we still can add try catch just in case. 

Basically what I'm saying is: if there is a possibility callback will throw it's better to add try catch that will suppress it and handle it without propagating the exception through additional callback or smth like that.

**Comment by MithrilMan** [comment]
after a call with @dangershony I'm now preparing another PR to implement a BackgroundWorkProvider that will be used to create and track AsyncQueue, AsyncLoop and AsyncDelegate

**Comment by dangershony** [comment]
>Suppress all exceptions (wrap callbacks in try catch), when exception occurs log it and then carry on or dispose the node based on component's importance.

Do we want to dispose the node in all cases?

**Comment by MithrilMan** [comment]
@noescape00 
> Basically what I'm saying is: if there is a possibility callback will throw it's better to add try catch that will suppress it and handle it without propagating the exception through additional callback or smth like that.

the problem is that you never know if someone change the code of the component and it start having a chance to throw, so is always better to have a fallback mechanism that trap unhandled exceptions (best would be to always try/catch the delegate code

Anyway now the implementation is moving toward a provider that wrap this logic itself
WIP PR: https://github.com/stratisproject/StratisBitcoinFullNode/pull/3455

**Pull request closed without merging**